src/devices/src/virtio/console/inorder_handler.rs

```diff
@@ -0,0 +1,112 @@
+use crate::virtio::SignalUsedQueue;
+use std::io::Write;
+use std::result;
+use virtio_console::console;
+use virtio_queue::{Queue, QueueStateOwnedT, QueueStateT};
+use vm_memory::GuestAddressSpace;
+
+#[derive(Debug)]
+pub enum Error {
+    GuestMemory(vm_memory::GuestMemoryError),
+    Queue(virtio_queue::Error),
+    Console(console::Error),
+}
+
+impl From<vm_memory::GuestMemoryError> for Error {
+    fn from(e: vm_memory::GuestMemoryError) -> Self {
+        Error::GuestMemory(e)
+    }
+}
+
+impl From<virtio_queue::Error> for Error {
+    fn from(e: virtio_queue::Error) -> Self {
+        Error::Queue(e)
+    }
+}
+
+impl From<console::Error> for Error {
+    fn from(e: console::Error) -> Self {
+        Error::Console(e)
+    }
+}
+
+pub struct InOrderQueueHandler<M: GuestAddressSpace, S: SignalUsedQueue, T: Write> {
+    pub driver_notify: S,
+    pub transmitq: Queue<M>,
+    pub receiveq: Queue<M>,
+    pub console: console::Console<T>,
+}
+
+impl<M, S, T> InOrderQueueHandler<M, S, T>
+where
+    M: GuestAddressSpace,
+    S: SignalUsedQueue,
+    T: Write,
+{
+    pub fn process_transmitq(&mut self) -> result::Result<(), Error> {
+        // To see why this is done in a loop, please look at the `Queue::enable_notification`
+        // comments in `virtio_queue`.
+        loop {
+            self.transmitq.disable_notification()?;
+
+            while let Some(mut chain) = self
+                .transmitq
+                .state
+                .pop_descriptor_chain(self.transmitq.mem.memory())
+            {
+                self.console.process_transmitq_chain(&mut chain)?;
+
+                self.transmitq.add_used(chain.head_index(), 0)?;
```

Comment:
I think it would be better to check if the driver needs a notification each time we add an entry in the used ring (like we do for the other devices in vmm-ref). There's a comment about that [here](https://github.com/rust-vmm/vm-virtio/blob/main/crates/virtio-queue/src/state.rs#L416) and documented [here](https://docs.oasis-open.org/virtio/virtio/v1.1/csprd01/virtio-v1.1-csprd01.html#x1-400007).